### PR TITLE
fix(ThreadMembersUpdate): Only emit added & removed thread members

### LINKED ISSUE #6513: Possibility of same old & new structure for `Client#threadMembersUpdate`
**Please describe the problem you are having in as much detail as possible:**
`<Client>.threadMembersUpdate` has the possibility of emitting the same `Collection` for both the `oldMembers` & `newMembers` parameters.

This is achieved by the following:
1. Join a thread
2. Restart your bot to ensure reproducibility. **Do not fetch active threads**!
3. Leave the thread

Upon leaving the thread, `oldMembers` and `newMembers` will emit the same `Collection`. Needless to say, this is quite... pointless! This is because the thread channel's members didn't have the member in question cached, ergo the same structures.

**Include a reproducible code sample here, if possible:**

```js
const client = new (require("discord.js").Client)({
  intents: [
    "GUILDS",
    "GUILD_MEMBERS"
  ]
});

client.once("ready", () => {
  const thread = client.channels.resolve("thread_id"); // Thread
  console.log(thread.members.cache); // Collection(0) [Map] {}
});

// Emitting from thread_id
client.on("threadMembersUpdate", (oldMembers, newMembers) => {
  console.log(oldMembers.equals(newMembers)); // true
});

client.login("token");
```

**Possible Fixes:**
I guess the simple answer is the make the `oldMembers` parameter nullable after these operations:
https://github.com/discordjs/discord.js/blob/70cc0295f833cd988ea627b37d20536f73e21630/src/client/actions/ThreadMembersUpdate.js#L11-L20
We could just check to see if the size is equal and return `null` if so. By sending out `null`, we are essentially saying that an event occurred that we did not have enough data about (the old cache was unreliable).

~~I guess another way to do it would be to emit a "fake" `ThreadMember` that we create solely for it to be removed here just to ensure that the event fires with an actual `ThreadMember`. I'm unsure if this is possible.~~

**Further details:**

- discord.js version: 13.2.0-dev
- Node.js version: 16.6.1
- Operating system: macOS Big Sur 11.5.2
- Priority this issue should have – please be realistic and elaborate if possible: Low

**Relevant client options:**

- partials: none
- gateway intents: GUILDS & GUILD_MEMBERS
- other: none
- I have also tested the issue on latest main; Commit hash: f39b597e3120701287e86acdb247b4815f693c91

### PR DESCRIPTION
**Pull request by Jiralite**
**Please describe the changes this PR makes and why it should be merged:**
As explained in #6513, the event may emit the **same** `oldMembers` and `newMembers` which was... not useful. It's not possible to tell what actually changed there - only that _something_ changed. This is due to discord.js always assuming the thread member leaving is cached.

I believe it'd be best to do nothing (like other events) if only uncached thread members leave. By doing this, the event does not fire as it would emit the same `Collection` for both parameters. We can utilise a new partial (`Partials.ThreadMember`) for people who still want that data. This solves that issue, but there is another.

discord.js emits these added & removed thread members alongside the existing thread member cache. Since Discord does not send the current thread members, this is susceptible to a caching issue where there may be some thread members in both payloads... or there may not. To find out what truly changed, you would have to perform operations on the received `Collection`s to find out the thread members that caused this gateway event to occur. Therefore, I removed this caching behaviour and aligned it to the API: only emitting the added & removed thread members.

Now, the event works like this:

```JavaScript
<Client>.on(Events.ThreadMembersUpdate, (thread, addedMembers, removedMembers) => {
  console.log(thread); // This is the thread where members got updated!
  
  if (addedMembers.size > 0) {
    // Only the members that were added are here.
  }

  if (removedMembers.size > 0) {
    // Only the members that were removed are here.
    if (removedMembers.first().partial) {
      // Uncached thread member left the thread.
      // This will be included with the new partial (`Partials.ThreadMember`).
    }
  }
});
```

The thread channel is being emitted too to avoid the user doing `addedMembers.first()?.thread ?? removedMembers.first().thread`. We already have the thread internally so it seemed reasonable to include it. Additionally, users can access the thread's member cache to do what they need as before.

Feedback is welcome and appreciated!

This resolves #6513.

**Status and versioning classification:**
- Code changes have been tested against the Discord API, or there are no code changes
- I know how to update typings and have done so, or typings don't need updating
- This PR includes breaking changes (methods removed or renamed, parameters moved or removed)

### CODE REVIEW

File: packages/discord.js/src/client/actions/ThreadMembersUpdate.js [hunk 1]
```diff
@@ -8,24 +9,35 @@ class ThreadMembersUpdateAction extends Action {
...
+        // Uncached thread member(s) left.
+        return {};
+      }
 
       /**
        * Emitted whenever members are added or removed from a thread. Requires `GUILD_MEMBERS` privileged intent
        * @event Client#threadMembersUpdate
-       * @param {Collection<Snowflake, ThreadMember>} oldMembers The members before the update
-       * @param {Collection<Snowflake, ThreadMember>} newMembers The members after the update
+       * @param {ThreadChannel} thread The thread where members got updated
```
Comment (ImRodry) [inline packages/discord.js/src/client/actions/ThreadMembersUpdate.js:36]: This should be the last parameter imo

File: packages/discord.js/src/client/actions/ThreadMembersUpdate.js [hunk 2]
```diff
@@ -8,24 +9,35 @@ class ThreadMembersUpdateAction extends Action {
...
+        (_addedMembers, addedMember) => _addedMembers.set(addedMember.user_id, thread.members._add(addedMember)),
+        addedMembers,
+      );
 
-      data.removed_member_ids?.forEach(memberId => {
-        thread.members.cache.delete(memberId);
-      });
+      data.removed_member_ids?.reduce((removedMembersIds, removedMembersId) => {
+        const cachedThreadMember = thread.members.cache.get(removedMembersId);
+        if (cachedThreadMember) removedMembersIds.set(cachedThreadMember.id, cachedThreadMember);
```
Comment (SpaceEEC) [inline packages/discord.js/src/client/actions/ThreadMembersUpdate.js:23]: I think something allowing you to know who left regardless of whether the thread member was cached would be nice.
(Partials or just a `null` value maybe?)

**Review by SpaceEEC — APPROVED** [review]

File: packages/discord.js/src/client/actions/ThreadMembersUpdate.js (hunk 2 above)
Comment (kyranet) [reply packages/discord.js/src/client/actions/ThreadMembersUpdate.js:23]: Yes, we should at very least give some data back, probably a PartialUser so we can read the IDs when iterating the maps with `.values()`

Comment (Jiralite) [reply packages/discord.js/src/client/actions/ThreadMembersUpdate.js]: I've gone ahead and created partial thread members. This should resolve the uncached behaviour here.

**Review by SpaceEEC — APPROVED** [review]

**Review by kyranet — APPROVED** [review]

**Comment by iCrawl** [comment]
This needs a rebase.

**Comment by iCrawl** [comment]
This needs a rebase.

**Review by iCrawl — APPROVED** [review]

**Pull request merged**